track_fil_state.go

```diff
@@ -0,0 +1,343 @@
+package pgchainbs
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/filecoin-project/go-bs-postgres-chainnotated/lib/synccid"
+	"github.com/filecoin-project/go-state-types/abi"
+	"github.com/ipfs/go-cid"
+
+	"github.com/jackc/pgx/v4"
+	"golang.org/x/xerrors"
+)
+
+// DestructuredFilTipSetData and DestructuredFilTipSetHeaderBlock represent the
+// pre-parsed contents of a Filecoin tipset structure. The indirection is
+// unavoidable as using lotus-defined types leads to import cycles.
+type DestructuredFilTipSetData struct {
+	Epoch                    abi.ChainEpoch
+	ParentWeight             string
+	ParentBaseFee            string
+	ParentStaterootCid       cid.Cid
+	ParentMessageReceiptsCid cid.Cid
+	ParentTipSetCids         []cid.Cid
+	HeaderBlocks             []DestructuredFilTipSetHeaderBlock
+	BeaconRoundAndData       [][]byte
+}
+type DestructuredFilTipSetHeaderBlock struct {
+	MinerActorID             uint64
+	HeaderCid                cid.Cid
+	MessagesCid              cid.Cid
+	TypedSignature           []byte
+	ForkSignalVarint         []byte
+	TicketProof              []byte
+	ElectionWincountAndProof []byte
+	WinpostTypesAndProof     [][]byte
+}
+
+// CurrentFilTipSetKeyBytes queries the configured instance namespace and
+// retrieves the epoch and ordered CIDs of the most-recently Visit()-ed tipset.
+func (dbbs *PgBlockstore) CurrentFilTipSetKey(ctx context.Context) ([]cid.Cid, abi.ChainEpoch, error) {
```

Comment:
might be nice to also expose the 'get tipset n behind current tipset' (the lookback view used by e.g. visor)

Reply:
@willscott: how is this?

```
// GetFilTipsetCurrent queries the configured instance namespace and retrieves
// the epoch and ordered CIDs of the most-recently Visit()-ed tipset.
func (dbbs *PgBlockstore) GetFilTipsetCurrent(ctx context.Context) ([]cid.Cid, abi.ChainEpoch, error) { ... }

// FindFilTipset queries the shared tipset list for the tipset formed from the
// provided chain header block CIDs. If the optional walkBackEpochs is larger
// than 0, return the parent at that distance from the initial given tipset. 
func (dbbs *PgBlockstore) FindFilTipset(ctx context.Context, tipsetCids []cid.Cid, walkBackEpochs abi.ChainEpoch) ([]cid.Cid, abi.ChainEpoch, error) { ... }
```

Reply:
that would be great

Reply:
@willscot switched this up a bit more: https://github.com/filecoin-project/go-bs-postgres-chainnotated/commit/cc21260af03ef4bfed6ba4b0c8a61fed8871a19f#diff-a7784a9261f9c1e3044236d29c9919f789463b57051b784afd90e02c7283bc92R42-R100

TLDR:
https://pkg.go.dev/github.com/filecoin-project/go-bs-postgres-chainnotated@v0.0.0-20210307170649-cc21260af03e#PgBlockstore.FindFilTipSet
+
https://pkg.go.dev/github.com/filecoin-project/go-bs-postgres-chainnotated@v0.0.0-20210307170649-cc21260af03e#PgBlockstore.GetFilTipSetHead